jms/src/main/scala/akka/stream/alpakka/jms/impl/JmsTxSourceStage.scala

```diff
@@ -42,48 +42,39 @@ private[jms] final class JmsTxSourceStage(settings: JmsConsumerSettings, destina
-                  try {
-                    val envelope = TxEnvelope(message, session)
-                    handleMessage.invoke(envelope)
-                    try {
-                      // JMS spec defines that commit/rollback must be done on the same thread.
-                      // While some JMS implementations work without this constraint, IBM MQ is
-                      // very strict about the spec and throws exceptions when called from a different thread.
-                      val action = Await.result(envelope.commitFuture, settings.ackTimeout)
-                      action()
-                    } catch {
-                      case _: TimeoutException =>
-                        val exception = new JmsTxAckTimeout(settings.ackTimeout)
-                        session.session.rollback()
-                        if (settings.failStreamOnAckTimeout) {
-                          handleError.invoke(exception)
-                        } else {
-                          log.warning(exception.getMessage)
-                        }
+    override protected def onSessionOpened(consumerSession: JmsConsumerSession): Unit =
+      consumerSession
+        .createConsumer(settings.selector)
+        .map { consumer =>
+          consumer.setMessageListener(new jms.MessageListener {
+            def onMessage(message: jms.Message): Unit = {
+              try {
+                val envelope = TxEnvelope(message, consumerSession)
+                handleMessage.invoke(envelope)
+                try {
+                  // JMS spec defines that commit/rollback must be done on the same thread.
+                  // While some JMS implementations work without this constraint, IBM MQ is
+                  // very strict about the spec and throws exceptions when called from a different thread.
+                  val action = Await.result(envelope.commitFuture, settings.ackTimeout)
+                  action()
+                } catch {
+                  case _: TimeoutException =>
+                    val exception = new JmsTxAckTimeout(settings.ackTimeout)
+                    consumerSession.session.rollback()
+                    if (settings.failStreamOnAckTimeout) {
+                      handleError.invoke(exception)
+                    } else {
+                      log.warning(exception.getMessage)
                     }
-                  } catch {
-                    case e: IllegalArgumentException => handleError.invoke(e) // Invalid envelope. Fail the stage.
-                    case e: jms.JMSException => handleError.invoke(e)
-                  }
-              })
+                }
+              } catch {
+                case e: IllegalArgumentException => handleError.invoke(e) // Invalid envelope, fail the stage
+                case e: jms.JMSException => handleError.invoke(e)
+              }
             }
-            .onComplete(sessionOpenedCB.invoke)
-
-        case _ =>
-          throw new IllegalArgumentException(
```

Comment:
Was wondering why the compiler wasn't warning this was unreachable... turns out that a `null` would be the only value which hits this... the new code would NPE on `createConsumer` which seems to be preferable (you go straight to "why was the argument null").

Scala 3 should complain about this case being unreachable (with the warning saying that it's only reachable in the null case, so recommend explicitly matching `null`).